### PR TITLE
feat: dial multiple peers in parallel

### PR DESCRIPTION
**Pull request by danisharora099**
## Problem

This PR addresses point 2 & point 3 of https://github.com/waku-org/js-waku/issues/1326

## Solution

- multiple discovered peers are dialed in parallel until max limit is reached
- connection to a bootstrap peer is dropped if more than >set bootstrap connections are formed

## Notes
- https://github.com/waku-org/js-waku/pull/1379 should be merged first

### CODE REVIEW

File: .cspell.json [hunk 1]
```diff
@@ -50,6 +50,8 @@
     "ihave",
     "ihaves",
     "ineed",
+    "initalised",
+    "initialisation",
```
Comment (fryorcraken) [inline .cspell.json:54]: wrong spelling.... (in US English I guess)

init**i**alized
initialization

File: packages/core/src/lib/connection_manager.ts [hunk 2]
```diff
@@ -60,6 +65,27 @@ export class ConnectionManager {
     this.run()
       .then(() => log(`Connection Manager is now running`))
       .catch((error) => log(`Unexpected error while running service`, error));
+
+    // libp2p emits `peer:discovery` events during its initialisation
+    // which means that before the ConnectionManager is initalised, some peers may have been discovered
```
Comment (fryorcraken) [inline packages/core/src/lib/connection_manager.ts:70]: shouldn't we initialize connection manager before starting the libp2p node? wouldn't that solve the issue with this race condition?

**Review by fryorcraken — COMMENTED** [review]
Looks good. How are you testing this?

File: packages/core/src/lib/connection_manager.ts [hunk 3]
```diff
@@ -60,6 +65,27 @@ export class ConnectionManager {
...
+    const peerInfos = await this.libp2pComponents.peerStore.all();
+    for (const peerInfo of peerInfos) {
+      if (
+        this.libp2pComponents
+          .getConnections()
+          .find((c) => c.remotePeer === peerInfo.id)
+      )
+        continue;
+
+      this.attemptDial(peerInfo.id).catch((error) => {
```
Comment (weboko) [inline packages/core/src/lib/connection_manager.ts:85]: We do not wait and `Promise` does not throw so no error can be handled outside of `dialPeerStorePeers`. 
To add logging I advice to map into an array of promises, wait for all and log an error for peers connection to which failed.

File: packages/core/src/lib/connection_manager.ts [hunk 4]
```diff
@@ -164,21 +217,50 @@ export class ConnectionManager {
...
 
+  private async attemptDial(peerId: PeerId): Promise<void> {
+    if (this.currentActiveDialCount >= this.options.maxParallelDials) {
+      this.pendingPeerDialQueue.push(peerId);
+      return;
+    }
+
+    if (!(await this.shouldDialPeer(peerId))) return;
+
+    this.dialPeer(peerId).catch((err) =>
```
Comment (weboko) [inline packages/core/src/lib/connection_manager.ts:228]: This method does not throw so better way would be to just `await`/`try` and remove `catch` where the method is used.

File: packages/core/src/lib/connection_manager.ts [hunk 5]
```diff
@@ -164,21 +217,50 @@ export class ConnectionManager {
...
+        peerId,
+        this.libp2pComponents.ping.bind(this)
+      );
+
+      const isBootstrap = (await this.getTagNamesForPeer(peerId)).includes(
+        Tags.BOOTSTRAP
+      );
+
+      if (isBootstrap) {
+        const bootstrapConnections = this.libp2pComponents
```
Comment (weboko) [inline packages/core/src/lib/connection_manager.ts:254]: Is it possible to determine `isBootstrap` by checking `conn.tags`? If so we can make `shouldDialPeer` sync and create a method to remove duplication of
```ts
this.libp2pComponents
        .getConnections()
        .filter((conn) => {
          conn.tags.find((name) => name === Tags.BOOTSTRAP);
        })```

**Review by fryorcraken — COMMENTED** [review]
Looks, good. Will let @weboko approve once his comments are tackled.

File: packages/core/src/lib/connection_manager.ts (hunk 5 above)
Comment (weboko) [reply packages/core/src/lib/connection_manager.ts:260]: Then lest resolve that comment in another PR or as follow-up

**Pull request merged**